crates/sui-core/src/transaction_input_loader.rs

```diff
@@ -0,0 +1,350 @@
+    #[instrument(level = "trace", skip_all)]
+    pub async fn read_objects_for_signing(
+        &self,
+        _tx_digest: &TransactionDigest,
+        input_object_kinds: &[InputObjectKind],
+        receiving_objects: &[ObjectRef],
+        protocol_config: &ProtocolConfig,
+        epoch_id: EpochId,
+    ) -> SuiResult<(InputObjects, ReceivingObjects)> {
+        fp_ensure!(
+            receiving_objects.len() + input_object_kinds.len()
+                <= protocol_config.max_input_objects() as usize,
+            UserInputError::SizeLimitExceeded {
+                limit: "maximum input and receiving objects in a transaction".to_string(),
+                value: protocol_config.max_input_objects().to_string()
+            }
+            .into()
+        );
+
+        let mut input_results = vec![None; input_object_kinds.len()];
+        let mut object_keys = Vec::with_capacity(input_object_kinds.len());
+        let mut fetch_indices = Vec::with_capacity(input_object_kinds.len());
+
+        for (i, kind) in input_object_kinds.iter().enumerate() {
+            match kind {
+                // Packages are loaded one at a time via the cache
+                InputObjectKind::MovePackage(id) => {
+                    let Some(package) = self.store.get_package_object(id)?.map(|o| o.into()) else {
+                        return Err(SuiError::from(kind.object_not_found_error()));
+                    };
+                    input_results[i] = Some(ObjectReadResult {
+                        input_object_kind: *kind,
+                        object: ObjectReadResultKind::Object(package),
+                    });
+                }
+                InputObjectKind::SharedMoveObject { id, .. } => match self.store.get_object(id)? {
+                    Some(object) => {
+                        input_results[i] = Some(ObjectReadResult::new(*kind, object.into()))
+                    }
+                    None => {
+                        if let Some((version, digest)) = self
+                            .store
+                            .get_last_shared_object_deletion_info(id, epoch_id)?
+                        {
+                            input_results[i] = Some(ObjectReadResult {
+                                input_object_kind: *kind,
+                                object: ObjectReadResultKind::DeletedSharedObject(version, digest),
+                            });
+                        } else {
+                            return Err(SuiError::from(kind.object_not_found_error()));
+                        }
+                    }
+                },
+                InputObjectKind::ImmOrOwnedMoveObject(objref) => {
+                    object_keys.push(ObjectKey::from(objref));
+                    fetch_indices.push(i);
+                }
+            }
+        }
+
+        let objects = self.store.multi_get_object_by_key(&object_keys)?;
+        for (index, object) in fetch_indices.into_iter().zip(objects.into_iter()) {
```

Comment:
Please add an assert_eq on the two sizes

Reply:
done